### ISSUE TITLE
"Portable" (no debug magic) mode using HERE-documents (<<'EOF')

### ISSUE THREAD

**Issue by Artoria2e5** [kristopolous/TickTick#44]
A lot of people are requesting ticktick for some other shell *x* where debug facilities may be nonexistent. A way to get that job done is to write a function that reads from its stdin, transpiles the code, and runs `eval`. People can then write blocks of JSON for that in [Here documents](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Here-Documents) and one-line here-strings.

For example, instead of the example, we can say:

```sh
#!/bin/bash

. ticktick.sh --no-magic

bob=Bob
tt <<'EOF'
  people = {
    "HR" : [
      "Alice",
      $bob,
      "Carol"
    ],
    "Sales": {
      "Gale": { "profits" : 1000 },
      "Harry": { "profits" : 500 }
    }
  }
EOF

function printEmployees() {
  echo
  echo "  The $(tt <<< people.Engineering.length()) Employees listed are:"

  for employee in $(tt <<< people.Engineering.items()); do
    printf "    - %s\n" ${!employee}
  done

  echo 
}

echo Base Assignment
tt <<< 'people.Engineering = [ "Darren", "Edith", "Frank" ]'
printEmployees

newPerson=Isaac
echo Pushed a new element by variable, $newPerson onto the array
tt <<< 'people.Engineering.push($newPerson)'
printEmployees

echo Shifted the first element off: $(tt <<< people.Engineering.shift())
printEmployees

echo Popped the last value off: $(tt <<< people.Engineering.pop())
printEmployees

echo Indexing an array, doing variable assignments

person0=$(tt <<< people.HR[0])
echo $person0 $(tt <<< people.HR[1])
```

(Note the quotes on EOF -- that prevents the shell from expanding `$bob` on its own prematurely.)

**Comment by kristopolous** [comment]
hey that's pretty clever. good work.

**Comment by kristopolous** [comment]
isn't <<< a bashism?

**Comment by kristopolous** [comment]
for instance, dash, from busybox doesn't have it:
 Redirection operators:
< > >| << >> <& >& <<- <>

**Comment by kristopolous** [comment]
nor does the old as dirt CSH
|| && | ^ & == != =~ !~ <= >=
< > << >> + - * / % ! ~ ( )

**Comment by kristopolous** [comment]
don't get me wrong, I love the hack but it doesn't fix the dash/csh/ksh/tcsh problem

**Comment by kristopolous** [comment]
I also don't really understand why jq and the more serious programs like that aren't alternatives ... maybe you can tell me ... what's your actual usecase?

**Comment by kristopolous** [comment]
I mean the real *real* **real** way to do this would be to just jack the interpreter all together ... essentially have a program say, `/usr/bin/crash` that does all the intepreter stuff and then do this 

```bash
#!/usr/bin/crash
... proceed as usual ...
```

**Comment by XLTechie** [comment]
> isn't <<< a bashism?

Yes.

http://mywiki.wooledge.org/HereDocument?action=show&redirect=HereString

**Comment by Artoria2e5** [comment]
Yeah, that's a bashism. `printf %s "bleh" | tt` can be used to go around it though. It might be available in some version of ksh, but I am not sure.

CSH is out of the question: the syntax isn't even in the same family of POSIX/Bourne shells.

**Comment by kristopolous** [comment]
I was in the space between sleep and awake this morning when I had more visions of the crash system ... essentially we'll need to either permanently do portable syntax (like '[' instead of '[[') or have a converter.  The '[[' is sooo much faster ... 

A bash-4 -> POSIX SH transpiler sounds like something that probably already exists in the world - I took enough of computer theory to know it's possible to completely make this since POSIX SH is tightly bound ... It's been around for 31 years I bet this exists. Let me look

**Comment by kristopolous** [comment]
hrmm ... "checkbashism" and "shellcheck" are the internet recommendations ... that's lame. I want to have a magic converter as part of a build script.

**Comment by kristopolous** [comment]
I dunno ... I think I'll just ask stackexchange ... https://unix.stackexchange.com/questions/588862/is-there-a-bash-to-posix-transpiler let's see what happens.

**Comment by kristopolous** [comment]
about 8 years ago, I was asked for busybox dash support in #22 and that should probably be the reference interpreter here. There's sadly quite a bit of hacks that would need to be done. bash's string substitution in a variable for instance `(${//})` doesn't seem to be there.

Unless I want this thing to crawl I think I'd have to rewrite chunks of it more heavily in awk and sed

**Comment by XLTechie** [comment]
Something I have wondered:

It claims to let you put JSON in *bash* scripts.
When did it decide it had to support every bash varient on the planet?

Let it be bash.

You never claimed it supported dash, busybox, korn, or aunt Fanny's shhhh.

**Comment by Artoria2e5** [comment]
One of the things for magic-less even in bash is usability in an interactive shell. That's #20.

**Comment by ewildgoose** [comment]
>  bash's string substitution in a variable for instance `(${//})` doesn't seem to be there.

I'm just checking my busybox shell right now and this works:

```
$ stringZ=abcABC123ABCabc; echo ${stringZ/abc/xyz}
xyzABC123ABCabc
```

Does this get us any further forward? The main thing we can't do with busybox is arrays.

I would love to get this working with my busybox systems so I can use it for simple config files. Nothing more than some fields with product type, version, etc. For various reasons it's helpful to have them in json.
